File: swift/Sources/EdgeVeda/PerfTrace.swift
/**
 * PerfTrace — JSONL frame-based performance trace logger.
 *
 * Records timestamped stage/value events grouped by frame ID.
 * Output format (one JSON object per line):
 *   {"frame_id": N, "ts_ms": T, "stage": "...", "value": V, ...}
 *
 * Usage:
 *   let trace = PerfTrace(fileURL: url)
 *   trace.record(stage: "decode", value: 12.5)
 *   trace.record(stage: "sample", value: 0.8, extra: ["top_k": 40])
 *   trace.nextFrame()
 *   trace.close()
 */

import Foundation

@available(iOS 15.0, macOS 12.0, *)
public actor PerfTrace {

    // MARK: - Properties

    private var frameId: Int = 0
    private let fileHandle: FileHandle?
    private let fileURL: URL?
    private var records: [[String: Any]] = []
    private var closed: Bool = false

    /// Session-level start time used to compute relative `ts_ms`.
    private let epochMs: Double

    // MARK: - Init

    /// Create a trace that writes JSONL to the given file URL.
    /// If `fileURL` is nil the trace accumulates records in memory
    /// (retrieve via `allRecords()`).
    public init(fileURL: URL? = nil) throws {
        self.fileURL = fileURL
        self.epochMs = Self.nowMs()

        if let url = fileURL {
            let dir = url.deletingLastPathComponent()
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            if !FileManager.default.fileExists(atPath: url.path) {
                FileManager.default.createFile(atPath: url.path, contents: nil)
            }
            self.fileHandle = try FileHandle(forWritingTo: url)
            self.fileHandle?.seekToEndOfFile()
        } else {
            self.fileHandle = nil
        }
    }

    // MARK: - Public API

    /// Record a trace event in the current frame.
    ///
    /// - Parameters:
    ///   - stage: Short label, e.g. "decode", "sample", "vision_preprocess".
    ///   - value: Numeric measurement (latency ms, token count, etc.).
    ///   - extra: Optional dictionary merged into the JSON line.
    public func record(stage: String, value: Double, extra: [String: Any]? = nil) {
        guard !closed else { return }

        var entry: [String: Any] = [
            "frame_id": frameId,
            "ts_ms": Self.nowMs() - epochMs,
            "stage": stage,
            "value": value,
        ]

        if let extra = extra {
            for (k, v) in extra {
                entry[k] = v
            }
        }

        records.append(entry)
        writeLineToFile(entry)
    }

    /// Advance to the next frame.
    public func nextFrame() {
        guard !closed else { return }
        frameId += 1
    }

    /// Current frame identifier.
    public func currentFrameId() -> Int {
        return frameId
    }

    /// Return all accumulated records (useful when no file URL was provided).
    public func allRecords() -> [[String: Any]] {
        return records
    }

    /// Flush and close the trace. No further recording is allowed.
    public func close() {
        guard !closed else { return }
        closed = true
        fileHandle?.closeFile()
    }

    /// Export all records as a single JSONL string.
    public func exportJSONL() -> String {
        return records.compactMap { entry in
            guard let data = try? JSONSerialization.data(withJSONObject: entry, options: [.sortedKeys]),
                  let line = String(data: data, encoding: .utf8) else {
                return nil
            }
            return line
        }.joined(separator: "\n")
    }

    // MARK: - Private Helpers

    private func writeLineToFile(_ entry: [String: Any]) {
        guard let fh = fileHandle else { return }
        guard let data = try? JSONSerialization.data(withJSONObject: entry, options: [.sortedKeys]),
              var line = String(data: data, encoding: .utf8) else {
            return
        }
        line.append("\n")
        if let lineData = line.data(using: .utf8) {
            fh.write(lineData)
        }
    }

    private static func nowMs() -> Double {
        return Date().timeIntervalSince1970 * 1000.0
    }
}

File: kotlin/src/main/kotlin/com/edgeveda/sdk/PerfTrace.kt
package com.edgeveda.sdk

import org.json.JSONObject
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * PerfTrace — JSONL frame-based performance trace logger.
 *
 * Records timestamped stage/value events grouped by frame ID.
 * Output format (one JSON object per line):
 *   {"frame_id": N, "ts_ms": T, "stage": "...", "value": V, ...}
 *
 * Usage:
 *   val trace = PerfTrace(file = File("trace.jsonl"))
 *   trace.record("decode", 12.5)
 *   trace.record("sample", 0.8, mapOf("top_k" to 40))
 *   trace.nextFrame()
 *   trace.close()
 */
class PerfTrace(
    file: File? = null,
) {
    private val frameId = AtomicInteger(0)
    private val closed = AtomicBoolean(false)
    private val epochMs: Double = nowMs()

    private val writer: BufferedWriter? = file?.let {
        it.parentFile?.mkdirs()
        BufferedWriter(FileWriter(it, /* append = */ true))
    }

    private val records = mutableListOf<Map<String, Any>>()
    private val lock = Any()

    // ── Public API ────────────────────────────────────────────────

    /**
     * Record a trace event in the current frame.
     *
     * @param stage Short label, e.g. "decode", "sample", "vision_preprocess".
     * @param value Numeric measurement (latency ms, token count, etc.).
     * @param extra Optional map merged into the JSON line.
     */
    fun record(stage: String, value: Double, extra: Map<String, Any>? = null) {
        if (closed.get()) return

        val entry = mutableMapOf<String, Any>(
            "frame_id" to frameId.get(),
            "ts_ms" to (nowMs() - epochMs),
            "stage" to stage,
            "value" to value,
        )
        extra?.let { entry.putAll(it) }

        synchronized(lock) {
            records.add(entry)
        }
        writeLineToFile(entry)
    }

    /** Advance to the next frame. */
    fun nextFrame() {
        if (closed.get()) return
        frameId.incrementAndGet()
    }

    /** Current frame identifier. */
    fun currentFrameId(): Int = frameId.get()

    /** Return a snapshot of all accumulated records. */
    fun allRecords(): List<Map<String, Any>> {
        synchronized(lock) {
            return records.toList()
        }
    }

    /** Flush and close the trace. No further recording is allowed. */
    fun close() {
        if (!closed.compareAndSet(false, true)) return
        try {
            writer?.flush()
            writer?.close()
        } catch (_: Exception) {
            // best-effort
        }
    }

    /** Export all records as a single JSONL string. */
    fun exportJSONL(): String {
        val snapshot = synchronized(lock) { records.toList() }
        return snapshot.joinToString("\n") { entry ->
            JSONObject(entry).toString()
        }
    }

    // ── Private Helpers ───────────────────────────────────────────

    private fun writeLineToFile(entry: Map<String, Any>) {
        val w = writer ?: return
        try {
            val line = JSONObject(entry).toString()
            synchronized(w) {
                w.write(line)
                w.newLine()
                w.flush()
            }
        } catch (_: Exception) {
            // best-effort — don't crash the caller
        }
    }

    private fun nowMs(): Double = System.currentTimeMillis().toDouble()
}